### PR TITLE
Remove the uses of split() and find() in the frontend

### PR DESCRIPTION
**Pull request by dengyuchenkit**
This PR removes string methods in ProtocolHandler:
- ProtocolHandler calls parser to check syntax error and get query type. Add tests to check that queries with syntax error or empty query are handled correctly. 
- Store a SQLStatementList inside Statement for replanning purpose.
- Fix the response when clients attempt to operate in a transaction which has already been aborted (such as due to SQL syntax error) but is not explicitly committed/rollbacked yet.

```
postgres=# begin;
BEGIN
postgres=# create tabel foo(id int);
syntax error at or near "tabel" at 8
postgres=# select * from foo;
current transaction is aborted, commands ignored until end of transaction block
postgres=# rollback;
ROLLBACK
```
- Fix PREPARE and EXECUTE query
```
postgres=# create table foo(id decimal);
CREATE TABLE
postgres=# prepare fooplan as insert into foo values($1);
PREPARE
postgres=# execute fooplan(1);
INSERT 0 1
postgres=# execute fooplan(1+1);
INSERT 0 1
postgres=# execute fooplan(SQRT(9.0));
INSERT 0 1
postgres=# select * from foo;
 id 
----
  1
  2
  3
(3 rows)
postgres=# 
```

### CODE REVIEW

**Comment by dengyuchenkit** [comment]
@hzxa21 I change the map to unordered_map and the shared_ptr back to unique_ptr. Could you please check if it's fine?

File: src/type/types.cpp [hunk 1]
```diff
@@ -549,6 +554,115 @@ std::ostream& operator<<(std::ostream& os, const StatementType& type) {
...
+    case QueryType::QUERY_ANALYZE: return "ANALYZE";
+    case QueryType::QUERY_RENAME: return "RENAME";
+    case QueryType::QUERY_PREPARE: return "PREPARE";
+    case QueryType::QUERY_EXECUTE: return "EXECUTE";
+    case QueryType::QUERY_SELECT: return "SELECT";
+    case QueryType::QUERY_OTHER: default: return "OTHER";
+  }
+}
+
+QueryType StringToQueryType(std::string str) {
```
Comment (chenboy) [inline src/type/types.cpp]: Since we only use the value of the string, it would be better if we use a `const std::string *` as the parameter to avoid copying. Also, seems like we will construct the hash map each time we call this function.

File: src/type/types.cpp [hunk 2]
```diff
@@ -549,6 +554,115 @@ std::ostream& operator<<(std::ostream& os, const StatementType& type) {
...
+  };
+  std::unordered_map<std::string, QueryType>::iterator it  = querytype_string_map.find(str);
+  if (it != querytype_string_map.end()) {
+    return it -> second;
+  } else {
+    return QueryType::QUERY_INVALID;
+  }
+}
+
+QueryType StatementTypeToQueryType(StatementType stmt_type, parser::SQLStatement* sql_stmt) {
```
Comment (chenboy) [inline src/type/types.cpp]: Seems like we will not modify sql_stmt in this function, please make it a `const parser::SQLStatement*`, also the hash map seems to be created each time we call this function.

File: src/include/common/statement.h [hunk 3]
```diff
@@ -81,6 +75,10 @@ class Statement : public Printable {
 
   const std::shared_ptr<planner::AbstractPlan>& GetPlanTree() const;
 
+  std::unique_ptr<parser::SQLStatementList>const& GetStmtParseTreeList() {return sql_stmt_list_;}
```
[the remark this answers is not in the document]
Comment (dengyuchenkit) [reply src/include/common/statement.h]: This method will be called when the statement lends its parse tree to the optimizer to generate a query plan. I think it may be better to keep this method now since the optimizer takes in a unique_ptr.

File: src/network/postgres_protocol_handler.cpp [hunk 4]
```diff
@@ -237,179 +235,132 @@ bool PostgresProtocolHandler::HardcodedExecuteFilter(QueryType query_type) {
...
+    traffic_cop_->AbortInvalidStmt();
+    std::string error_message = e.what();
+    SendErrorResponse(
+        {{NetworkMessageType::HUMAN_READABLE_ERROR, e.what()}});
+    SendReadyForQuery(NetworkTransactionStateType::IDLE);
+    return ProcessResult::COMPLETE;
+  }
+  // If the query contains zero sql statements, for example ";;;"
+  if (sql_stmt_list->GetNumStatements() == 0) {
+    SendEmptyQueryResponse();
```
[the remark this answers is not in the document]
Comment (dengyuchenkit) [reply src/network/postgres_protocol_handler.cpp]: Yes, thanks!

File: src/network/postgres_protocol_handler.cpp [hunk 5]
```diff
@@ -237,179 +235,132 @@ bool PostgresProtocolHandler::HardcodedExecuteFilter(QueryType query_type) {
...
-                                           results_, rows_affected_, error_message_, thread_id);
-        if (traffic_cop_->is_queuing_) {
-          return ProcessResult::PROCESSING;
-        }
-        ExecQueryMessageGetResult(status);
+      query_ = statement_->GetQueryString();
+      query_type_ = statement_->GetQueryType();
+      std::vector<int> result_format(statement_->GetTupleDescriptor().size(), 0);
+      result_format_ = result_format;
+      //TODO: We assume it's only constant value expression
```
[the remark this answers is not in the document]
Comment (dengyuchenkit) [reply src/network/postgres_protocol_handler.cpp]: Okay, I will do that.

File: src/type/types.cpp (hunk 2 above)
Comment (dengyuchenkit) [reply src/type/types.cpp]: Fixed

File: src/type/types.cpp (hunk 1 above)
Comment (dengyuchenkit) [reply src/type/types.cpp]: Fixed

File: src/network/postgres_protocol_handler.cpp [hunk 6]
```diff
@@ -237,179 +235,132 @@ bool PostgresProtocolHandler::HardcodedExecuteFilter(QueryType query_type) {
...
+      for (std::unique_ptr<expression::AbstractExpression>& param : exec_stmt->parameters) {
```
[the remark this answers is not in the document]
Comment (dengyuchenkit) [reply src/network/postgres_protocol_handler.cpp]: Hi just to clarify first, the exec_stmt->parameters is a vector of unique pointers. I suppose it's more convenient to iterate by the reference of the unique pointers. Could you explain more if you still think it's better to use the raw pointer here? Thanks

File: src/network/postgres_protocol_handler.cpp [hunk 7]
```diff
@@ -237,179 +235,132 @@ bool PostgresProtocolHandler::HardcodedExecuteFilter(QueryType query_type) {
...
+      // There used to be code here that would invoke this method passing
+      // in NetworkMessageType::READY_FOR_QUERY as the argument. But when
+      // I switched to strong types, this obviously doesn't work. So I
+      // switched it to be NetworkTransactionStateType::IDLE. I don't know
+      // we just don't always send back the internal txn state?
+      SendReadyForQuery(NetworkTransactionStateType::IDLE);
+      return ProcessResult::COMPLETE;
+    };
+    case QueryType::QUERY_EXECUTE: {
+      parser::ExecuteStatement* exec_stmt = (parser::ExecuteStatement*)sql_stmt.get();
```
[the remark this answers is not in the document]
Comment (dengyuchenkit) [reply src/network/postgres_protocol_handler.cpp]: I change it to dynamic_cast. Right now this kind of casting (dynamic_cast or static_cast) exist in a large number of codes. I think maybe right now it's fine to keep it in the same way, for consistency.

File: test/network/exception_test.cpp [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+      if (s) {
+        LOG_INFO("Invalid Drop Query: %s", e.base().what());
+        exception_count += 1;
+      }
+    }
+
+    // CREATE query
+    pqxx::work txn2(C);
+    try {
+      // TODO(Yuchen): If the query is 'CREATE TABLE employee(id ITN)', postgres parser will generate a valid parse tree
```
Comment (dengyuchenkit) [inline test/network/exception_test.cpp]: @haojin2 I am wondering, there may be a memory leak in the parser. Could you have a look it?
It happens when issuing query _CREATE TABLE employee(id ITN)_. It is invalid since the typename _ITN_ is wrong. The postgres parser returns a valid parse tree. But Peloton will catch the error when doing node transform. The memory block allocated for 'result' object in CreateTransform() is not freed. I see that ListTransform() will free the SQLStatementList. The memory allocated in CreateTransform() may still not be freed since the CreateStatement is not added into the SQLStatementList yet. I am thinking maybe we need some changes in the exception handling. Could you check that? Thanks

File: src/traffic_cop/traffic_cop.cpp [hunk 9]
```diff
@@ -281,15 +275,23 @@ void TrafficCop::ExecuteStatementPlanGetResult() {
...
+ * Prepare a statement based on parse tree. Begin a transaction if necessary.
+ * If the query is not issued in a transaction (when the query type is not 'BEGIN'
+ * and the tcop_txn_state is empty), Peloton will wrap it in a transaction and
+ * begin a transaction here. It is a single_stmt transaction.
+ * Otherwise, the query is issued inside a transaction by the client. It is a multi_stmt transaction.
+ */
 std::shared_ptr<Statement> TrafficCop::PrepareStatement(
-    const std::string &statement_name, const std::string &query_string,
-    UNUSED_ATTRIBUTE std::string &error_message,
+    const std::string &stmt_name, const std::string &query_string,
```
Comment (ChTimTsubasa) [inline src/traffic_cop/traffic_cop.cpp]: I am wondering if we still need the query string here if we already do the parsing in the protocol handler?

File: src/traffic_cop/traffic_cop.cpp [hunk 10]
```diff
@@ -210,35 +212,27 @@ executor::ExecuteResult TrafficCop::ExecuteStatementPlan(
...
     threadpool::MonoQueuePool::GetInstance().SubmitTask(ExecutePlanWrapper, arg, task_callback_, task_callback_arg_);
     LOG_TRACE("Submit Task into MonoQueuePool");
-
     is_queuing_ = true;
     return p_status_;
-
   } else {
-    // otherwise, we have already aborted
-    p_status_.m_result = ResultType::ABORTED;
+    // If the transaction state is ABORTED, the transaction should be aborted
```
[the remark this answers is not in the document]
Comment (dengyuchenkit) [reply src/traffic_cop/traffic_cop.cpp]: It is fixed. It is a comment.

**Comment by dengyuchenkit** [comment]
@ChTimTsubasa @chenboy Would you mind taking a second round reviewing this? It is ready to be reviewed again

File: src/include/traffic_cop/traffic_cop.h [hunk 11]
```diff
@@ -129,15 +124,14 @@ class TrafficCop {
   std::unique_ptr<optimizer::AbstractOptimizer> optimizer_;
 
   // flag of single statement txn
-  bool single_statement_txn_;
```
Comment (ChTimTsubasa) [inline src/include/traffic_cop/traffic_cop.h]: Set this in the constructor.

File: src/include/type/types.h [hunk 12]
```diff
@@ -12,7 +12,7 @@
 
 #pragma once
 
-#include <bitset>
```
Comment (ChTimTsubasa) [inline src/include/type/types.h]: Why is this blank missing?

File: src/traffic_cop/traffic_cop.cpp (hunk 9 above)
Comment (ChTimTsubasa) [reply src/traffic_cop/traffic_cop.cpp]: I think Haoxiang has a PR that fix this problem. Just keep it in this PR.

File: test/network/exception_test.cpp (hunk 8 above)
Comment (dengyuchenkit) [reply test/network/exception_test.cpp]: Fixed in #979

**Comment by dengyuchenkit** [comment]
@apavlo The problem I mentioned during the meeting is that, in ProtocolHandler when processing EXECUTE query, it does not call optimizer since it doesn't need a query plan. After talking to @chenboy , we realize that it results in the fact that we didn't call binder to check the params and evaluate the expression. The way of directly casting it to Const Expression is incorrect. Will check it

**Comment by dengyuchenkit** [comment]
@tli2 @haojin2 Could you help me review this PR? I think Tianyi had a pass over the ProtocolHandler and TrafficCop part. There're some changes in parser and binder. It will be great if @haojin2 can also take a look at it. Thanks

**Comment by haojin2** [comment]
LGTM so far.

**Comment by dengyuchenkit** [comment]
@apavlo This PR should be ready now.

**Comment by apavlo** [comment]
@dengyuchenkit I will merge this after the build passes.
